src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2372,46 +2372,152 @@ private void SelectVoice(string voice)
+                    var vEnd = commandList.FindLastIndex(x => x == "/Loop>") - 1;
+                    for (int i=0; i < count; i++)
                     {
-                        Enums.Keyboards keyboardEnum;
-                        bool vReturn = (vStep.StartsWith("KeyboardAndReturn>"));
-                        int vIndex = (vStep.StartsWith("KeyboardAndReturn>")) ? 18 : 9;
-                        if (System.Enum.TryParse(vStep.Substring(vIndex), out keyboardEnum))
-                        {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
-                        }
-                        else
-                        {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(vStep.Substring(vIndex), vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
-                        }
+                        var loopList = commandList.GetRange(1, vEnd);
+                        StepListCommands(singleKeyValue, multiKeySelection, loopList);
+                    }
+                    commandList.RemoveRange(0, vEnd);
+                }
+                else if (commandList.First().StartsWith("Action>") &&
+                    (Enum.TryParse<FunctionKeys>(commandList.First().Substring(7), out FunctionKeys result)))
+                {
+                    keyValue = new KeyValue(result);
+                    KeySelectionResult(keyValue, multiKeySelection);
+                }
+                else if (commandList.First().StartsWith("Keyboard>") || commandList.First().StartsWith("KeyboardAndReturn>"))
+                {
+                    Enums.Keyboards keyboardEnum;
+                    bool vReturn = (commandList.First().StartsWith("KeyboardAndReturn>"));
+                    int vIndex = (commandList.First().StartsWith("KeyboardAndReturn>")) ? 18 : 9;
+                    if (System.Enum.TryParse(commandList.First().Substring(vIndex), out keyboardEnum))
+                    {
+                        keyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
+                    }
+                    else
+                    {
+                        keyValue = new ChangeKeyboardKeyValue(commandList.First().Substring(vIndex), vReturn);
+                    }
+                    KeySelectionResult(keyValue, multiKeySelection);
+                }
+                else if (commandList.First().StartsWith("KeyDown>"))
+                {
+                    keyValue = new KeyValue(commandList.First().Substring(8));
+                    int delayMs = 0;
+                    keyboardOutputService.ProcessSingleKeyPress(keyValue.String, KeyPressKeyValue.KeyPressType.Press, delayMs);
+                    if (keyStateService.KeyDownStates[singleKeyValue].Value != KeyDownStates.Down)
```

Comment:
I've found that the keyStateService can be out of sync from what Windows recognizes as the virtual state of a key.